feat: add program_lib property to Transpiler class

yasuhito opened the pull request:
## ✍ Description

Add `program_lib` property to the `Transpiler` class to explicitly specify which program format each transpiler natively handles. This change improves the program conversion logic by eliminating unnecessary conversions when a transpiler can directly handle a specific format.

This enhancement eliminates redundant program conversions (such as OpenQASM3 to OpenQASM3 when using OuquTpTranspiler), makes the transpiler's native format explicit in its interface, and simplifies program converter management by removing unnecessary converters.

### Changes

- Add abstract `program_lib` property to `Transpiler` class
- Implement `program_lib` property in concrete transpiler classes:
  - `QiskitTranspiler`: returns "qiskit"
  - `OuquTpTranspiler`: returns "openqasm3"
- Update `TranspilerDispatcher` to use `program_lib` for determining whether program format conversion is necessary

snuffkin reviewed (changes requested):
The design policy for `xxx_lib` is as follows:
Instead of specifying the exact name of `xxx_lib` in each component, it will either be defined by the user or provided as a built-in component by `Tranqu` class.

With that in mind, could you avoid hardcoding `program_lib` in Transpiler?
Could you add `program_lib` as an argument to the Transpiler's constructor?

yasuhito commented:
I have modified the code based on the design policy feedback. Here are the changes:

**Modified the Transpiler base class**:

```python
class Transpiler(ABC):
    def __init__(self, program_lib: str) -> None:
        self._program_lib = program_lib

    @property
    def program_lib(self) -> str:
        return self._program_lib
```

- Changed constructor to accept program_lib as a required parameter
- Exposed program_lib as a property


**Updated Transpiler implementations**:

QiskitTranspiler:
```python
    def __init__(self, program_lib: str) -> None:
        super().__init__(program_lib)
        self._stats_extractor = QiskitStatsExtractor()
        self._layout_mapper = QiskitLayoutMapper()
```

OuquTpTranspiler:
```python
    def __init__(self, program_lib: str) -> None:
        super().__init__(program_lib)
        self._ouqu_tp = OuquTp()
        self._qiskit_stats_extractor = QiskitStatsExtractor()
        self._layout_mapper = QiskitLayoutMapper()
```

Key points of the modification:
- Removed hardcoded program_lib values from transpiler implementations
- Now program_lib can be explicitly specified when registering transpilers in the Tranqu class

This modification aligns with the design policy where program_lib specification is handled by either the user or the Tranqu class rather than being hardcoded in individual components.

snuffkin reviewed (approved):
LGTM!
